### ISSUE TITLE
Import scala's `require` function to Akka.Util

### ISSUE THREAD

**Issue by Aaronontheweb** [akkadotnet/akka.net#570]
In many places within the original Akka codebase, the development team enforces code contracts using a simple scala function - `require`:

```scala
private[testkit] lazy val myself: RoleName = {
    require(_roles.size > MultiNodeSpec.selfIndex, "not enough roles declared for this test")
    _roles(MultiNodeSpec.selfIndex)
  }
```

See https://github.com/akka/akka/blob/master/akka-multi-node-testkit/src/main/scala/akka/remote/testkit/MultiNodeSpec.scala#L96

`require` seems pretty similar to `Debug.Assert` in that it throws an exception when a boolean expression evaluates to false. But I'm sure there are some nuances to it.

We've been hacking around the lack of this call inside tons of constructors in various parts within Akka.NET - I think it's time we added this to the codebase for parity purposes.

Please add this class to `Akka.Util` inside the code Akka library.

**Comment by rogeralsing** [comment]
This leads to another q, should we use code contracts?

Skickat från min iPhone

**Comment by Aaronontheweb** [comment]
I think using the specific "code contracts" .NET library that recently got OSSed is overkill, but I think we should use the concept itself liberally.

**Comment by alex-bogomaz** [comment]
There is one place in the codebase (DeadlineFailureDetector.cs) where explicit "throw new Ex.." is replaced with the function like ```require```:  
```csharp 
Helios.Util.Guard.True(bool boolean, string errorMessage)
```

```csharp
public DeadlineFailureDetector(TimeSpan acceptableHeartbeatPause, 
        {
            //.....
            Guard.True(_acceptableHeartbeatPause > TimeSpan.Zero, 
                       "acceptable-heartbeat-pause must be greater than zero");
        }
```
So we may say that we already have such function.
On the other hand scala's implementation of require is straightforward and matches the implementation of Guard.True (see https://github.com/scala/scala/blob/2.11.x/src/library/scala/Predef.scala#L205)
```scala
def require(requirement: Boolean, message: => Any) {
    if (!requirement)
      throw new IllegalArgumentException("requirement failed: "+ message)
  }
```

**Comment by Aaronontheweb** [comment]
great catch, @alex-bogomaz !

So in order to introduce this capability to Akka Core, we'd have to add a dependency on Helios - which I'd like to avoid in order to keep things simple.

IMHO, I think we should just clone Helios` `Guard` function into Akka.NET and add that to Akka.Util.

**Comment by avoxm** [comment]
@Aaronontheweb do you think we need to have everything ported from Helios ? Or this would be enough /

    /// <summary>
    /// Contains static methods for representing program contracts.
    /// </summary>
    public static class Guard
    {
        /// <summary>
        /// Checks for a condition;if the condition is false throws <see cref="System.ArgumentException"/> exception.
        /// </summary>
        /// <param name="condition">The conditional expression to test.</param>
        /// <param name="errorMessage">Exception message.</param>
        public static void Assert(bool condition, string errorMessage = "Expression should be true, but was false")
        {
            if (!condition)
                throw new ArgumentException(errorMessage);
        }
    }

I think we can gradually add only the staff we will need along the way.

Any ideas ?

**Comment by Aaronontheweb** [comment]
@avoxm I think that's perfect.

Helios has a bunch of its own Utils, similar to Akka.NET's Util space. Borrowing from those is fine as needed. We should only really add a Helios dependency to a module when we need Helios' socket capabilities.

**Comment by avoxm** [comment]
Ok, cool. I'll send PR and check #571

**Comment by Aaronontheweb** [comment]
Closed as of #571

**Comment by Aaronontheweb** [comment]
Err, closed as of pull request #578
